agent/src/beerocks/slave/agent_db.h

```diff
@@ -0,0 +1,76 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2019-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "task_pool.h"
+
+#include <iostream>
+#include <memory>
+#include <mutex>
+
+using namespace beerocks_message;
+using namespace son;
+using namespace beerocks;
+
+/**
+ * @brief AgentDB is a class used by all main Agent threads on the Agent process to store the Agent
+ * common data.
+ * The class is implemented using singleton design pattern.
+ * It is thread safe and being locked and released automatically.
+ * 
+ * How to use:
+ * 
+ * auto db = AgentDB::get();            // Get DB and automatically lock it. The lock will be
+ *                                      // Release when 'db' will be destructed.
+ * 
+ * db->foo = 42;                        // Change database member 'foo' to value 42
+ * db->bar = std::string("prpl")        // Set another database member
+ * 
+ * AgentDB::get()->foo = 44;            // Set database member directly.
+ * auto my_foo = AgentDB::get()->foo;   // Get a copy of database member.
+ * 
+ * auto &foo = AgentDB::get()->foo;     // Unsafe! Don't do it! The database will be unlocked,
+ *                                      // but the reference will remain.
+ */
+class AgentDB {
+public:
+    class SafeDB {
+    public:
+        SafeDB(AgentDB &db) : m_db(db) { m_db.db_lock(); }
```

Comment:
I'm don't understand why would we ever use std::move() on it? We can pass it as a reference to a sub-function and when it finishes the object is destructed.
Can you please give an example?

Reply:
It would be to allow inclusion in a super-structure. However, it's not a use case that is needed at the moment, and if we need it, we can still add the move constructor. So my bad.